org.eclipse.lsp4e/src/org/eclipse/lsp4e/DocumentContentSynchronizer.java

```diff
@@ -72,7 +72,10 @@ final class DocumentContentSynchronizer implements IDocumentListener {
 	private int version = 0;
 	private DidChangeTextDocumentParams changeParams;
 	private long modificationStamp;
-	private final AtomicReference<CompletableFuture<LanguageServer>> lastChangeFuture;
+	private CompletableFuture<LanguageServer> lastChangeFuture;
+	private final CompletableFuture<LanguageServer> didOpenFuture;
+	private final Object guard = new Object();
+	private long documentModificationStamp = 0L;
```

Comment:
AIUI `synchronised` is as strong as `volatile` in terms of its memory model guarantees, so that should happen automatically provided it is only accessed from the `synchronized` blocks. See https://jenkov.com/tutorials/java-concurrency/synchronized.html for example.